lib/core/src/control/mouse_manager.rs

```diff
@@ -0,0 +1,338 @@
+use crate::display::render::css3d::DOMContainer;
+use crate::system::web::dyn_into;
+use crate::system::web::Result;
+use crate::system::web::Error;
+use crate::system::web::ignore_context_menu;
+
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use web_sys::MouseEvent;
+use web_sys::WheelEvent;
+use web_sys::EventTarget;
+use web_sys::AddEventListenerOptions;
+use js_sys::Function;
+use nalgebra::Vector2;
+use std::rc::Rc;
+use std::cell::RefCell;
+
+
+
+// ==========================
+// === add_callback macro ===
+// ==========================
+
+macro_rules! add_callback {
+    ($name:ident, $event_type:ident, $target:literal) => { paste::item! {
+        pub fn $name
+        <F:[<$event_type Callback>]>(&mut self, mut f:F) -> Result<MouseEventListener> {
+            let data = Rc::downgrade(&self.data);
+            let closure = move |event:MouseEvent| {
+                if let Some(data) = data.upgrade() {
+                    f([<$event_type Event>]::from(event, &data));
+                }
+            };
+            add_mouse_event(&self.data.target(), $target, closure)
+        }
+    } };
+}
```

Comment:
Please define the macro close to its usage point. Its hard to understand how it is used. Plus please provide docuementation to it - describing how to use it and what is the output of its usage (see docs to other macros).